File: ts/src/operations.ts
import { TodoError } from "./todo-error.js";
import { Operation, Todo, Todos } from "./types";

function validateId(todos: Todos, id: string, op: Operation): number {
  if (!Number.isInteger(+id))
    throw new TodoError("delete", "id must be a valid positive integer.");

  if (!todos.todoList[+id])
    throw new TodoError(op, `cannot find todo with id: ${id}`);

  return +id;
}

export function createTodo(todos: Todos, args: string[]): void {
  const [todo] = args;
  const newId = todos.nextId;
  todos.todoList[newId] = { isCompleted: false, todo };
  console.log(`Created todo "${todo}" with id: ${todos.nextId++}.`);
}

export function editTodo(todos: Todos, args: string[]): void {
  const [id, todo] = args;
  const todoId = validateId(todos, id, "edit");
  todos.todoList[todoId] = {
    ...todos.todoList[todoId],
    todo,
  };
  console.log(`Todo id: ${id} successfully edited.`);
}

export function viewTodo(todos: Todos, args: string[]): Todo[] {
  const [cmd] = args;

  if (cmd === "all") {
    const todoList: Todo[] = [];
    for (const key in todos.todoList) {
      const todo = todos.todoList[key];
      console.log(`${key}:  [${todo.isCompleted ? "✓" : " "}] ${todo.todo}`);
      todoList.push({ isCompleted: todo.isCompleted, todo: todo.todo });
    }
    return todoList;
  } else {
    const todoId = validateId(todos, cmd, "view");
    const todo = todos.todoList[todoId];
    console.log(`${todoId}:  [${todo.isCompleted ? "✓" : " "}] ${todo.todo}`);
    return [{ isCompleted: todo.isCompleted, todo: todo.todo }];
  }
}

export function updateTodo(
  todos: Todos,
  args: string[],
  isFinished: boolean
): void {
  const [id] = args;
  const todoId = validateId(todos, id, "finish");
  todos.todoList[todoId] = {
    ...todos.todoList[todoId],
    isCompleted: isFinished,
  };
}

export function deleteTodo(todos: Todos, args: string[]): void {
  const [id] = args;
  const todoId = validateId(todos, id, "delete");
  const deletedTodo = todos.todoList[todoId];
  delete todos.todoList[todoId];
  console.log(`Todo id: ${id} successfully deleted.`);
}

File: py/operations.py
from todos import *

def createTodo(todos: Todos, argv: list[str]):
  if len(argv) < 3:
    print("Usage: python app.py arg1 arg2 arg3")
    return
  
  todoStr = argv[2]
  newTodo = Todo(isCompleted=False, todo=todoStr)
  todos.todoList[todos.nextId] = newTodo.toJson()
  todos.nextId += 1

def viewTodo(todos: Todos, argv: list[str]):
  if len(argv) < 3:
    print("Usage: python app.py arg1 arg2 arg3")
    return
  
  cmd = argv[2]
  if cmd == "all":
    for key, val in todos.todoList.items():
      status = "✓" if val["isCompleted"] else " "
      todo = val["todo"]
      print(f"{key}: [{status}] {todo}")
  else:
    if not cmd.isdigit():
      print(f"Specified key '{cmd}' is not a positive integer.")
      return
    
    if cmd not in todos.todoList:
      print(f"Todo with key '{cmd}' does not exist in your todos.")
      return
    
    todo = todos.todoList[cmd]
    status = "✓" if todo["isCompleted"] else " "
    todoStr = todo["todo"]
    print(f"{cmd}: [{status}] {todoStr}")
      
def editTodo(todos: Todos, argv: list[str]):
  if len(argv) < 4:
    print("Usage: python app.py arg1 arg2 arg3")
    return
  
  key = argv[2]
  todoStr = argv[3]
  
  if not key.isdigit():
    print(f"Specified key '{key}' is not a positive integer.")
    return
    
  if key not in todos.todoList:
    print(f"Todo with key '{key}' does not exist in your todos.")
    return
  
  todos.todoList[key]["todo"] = todoStr
  
def deleteTodo():
  print("Delete")
  
def updateTodo(isFinished: bool):
  print(isFinished)
